File: Study_Problem_Solving/[2021-pra]4-1.cpp
#include <stdio.h>
#include <deque>
#include <vector>
using namespace std;

int main()
{

    int n;
    scanf("%d", &n);

    int top[500001];
    int answer[500001];
    deque<vector<int>> stack;

    for (int i = 0; i < n; i++)
    {
        int temp;
        scanf("%d", &temp);
        top[i] = temp;
    }

    for (int i = 0; i <n; i++){
        while (!stack.empty()){
            if (stack.back()[0] >= top[i]){
                answer[i] = stack.back()[1] + 1;
                vector<int> temp = {top[i], i};
                stack.push_back(temp);
                break;
            }
            else{
                stack.pop_back();
            }
        }

        if (stack.empty()){
            answer[i] = 0;
            vector<int> temp = {top[i], i};
            stack.push_back(temp);
        }
    }


    for (int i = 0; i < n; i++)
    {
        printf("%d ", answer[i]);
    }
    return 0;
}

File: Study_Problem_Solving/[2021-pra]4-1.py
from collections import deque
import sys
input = sys.stdin.readline

N = int(input())
top = deque(list(map(int, input().split())))
stack = []
answer = []
i = 0
while top:
    value, index = top.popleft(), i

    while stack:
        if stack[-1][0] > value:
            answer.append(stack[-1][1] + 1)
            stack.append([value, index])
            break
        else:
            stack.pop()

    if len(stack) == 0:
        answer.append(0)
        stack.append([value, index])
    i += 1

for element in answer:
    print(element, end = " ")
